### PR TITLE
feature(icons): standardized API for handling entity icons

### PR DESCRIPTION
**Pull request by hypeJunction**
**feature(icons): standardized API for handling entity icons**

Adds API for working with entity icons:
`elgg_get_icon_sizes()` to retreive icon_sizes config
`elgg_get_entity_icon()` to retrieve an icon as an ElggFile object
`elgg_create_entity_icons()` to generate entity icons
`elgg_clear_entity_icons()` to clear entity icons

All icons now have a predicatable location in `/icons/$size.jpg` relative to entity's filestore directory.

Legacy icons are moved to the new location whenever requested. A symlink is established from the old location to the new location.

Icons sizes are now configurable via `"entity:icon:sizes, $entity_type"` hook.

**feature(embed): adds embed-icon page handler**

Adds a handler to serve embedded icons. Adds `elgg_get_embed_url()` to generate embed URLs
- [x] Add docs
- [ ] Update `ElggDiskFilestore::getSize` to include all entities owned by container
- [x] Symlink old icon location to the new one

### CODE REVIEW

**Comment by hypeJunction** [comment]
To consider #6181

File: engine/lib/filestore.php [hunk 1]
```diff
@@ -602,6 +602,205 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = ''
...
+		$icon->owner_guid = $entity->owner_guid;
+		$icon->setFilename("groups/{$entity->guid}{$size}.jpg");
+	}
+
+	if ($icon->exists()) {
+		return $icon;
+	}
+
+	$icon->owner_guid = $entity->guid;
+	$icon->setFilename("icons/{$size}.jpg");
```
Comment (hypeJunction) [inline engine/lib/filestore.php]: For BC, we can add a symlink from profile/groups to the new location.

**Comment by juho-jaakkola** [comment]
Looks good so far!

File: docs/admin/upgrading.rst [hunk 2]
```diff
@@ -53,6 +53,7 @@ Group avatars are now served via ``serve-file`` handler. Plugins should start us
 
  * ``groupicon`` page handler (``groups_icon_handler()``) has been deprecated
  * ``/mod/groups/icon.php`` file has been deprecated
+ * ``groups_set_icon_url()`` no longer in use as callback for ``entity:icon:url, group`` hook
```
Comment (hypeJunction) [inline docs/admin/upgrading.rst]: Bad idea. Need to bring it back. It sets the default icon...

**Comment by hypeJunction** [comment]
This now works with all entity types.

File: docs/guides/upgrading.rst [hunk 3]
```diff
@@ -36,6 +38,56 @@ Added ``elgg/lightbox`` module
...
+      }
+
+      // proceed to determining the URL for icons generated the old way
+   }
+
+
+Added handlers for file embeds
+------------------------------
+
+ * ``elgg_get_embed_url()`` - can be used to return an embed URL for a file's thumbnail
```
Comment (hypeJunction) [inline docs/guides/upgrading.rst]: Change to `embed-icon`

File: engine/classes/ElggEntity.php [hunk 4]
```diff
@@ -274,6 +274,12 @@ public function __get($name) {
 			return 2;
 		}
 
+		if (in_array($name, array('thumbnail', 'smallthumb', 'largethumb'))) {
```
Comment (hypeJunction) [inline engine/classes/ElggEntity.php]: Is there a way to deprecate this without generating a ton of warning when we access these values?

File: engine/lib/filestore.php [hunk 5]
```diff
@@ -481,12 +445,52 @@ function _elgg_filestore_init() {
...
+
+/**
+ * Reset icon URLs if access_id has changed
+ *
+ * @param string     $event  "update:after"
+ * @param string     $type   "object"|"group"
+ * @param ElggObject $entity Entity
+ * @return void
+ */
+function _elgg_filestore_touch_icons($event, $type, $entity) {
```
Comment (hypeJunction) [inline engine/lib/filestore.php]: make `@private`

**Comment by hypeJunction** [comment]
I guess I will need to break this down into multiple PRs, but let's first agree on the strategy:

1. Whenever an icon is requested, we first check the old location for known types (user, group, file)
2. If the old icon exists, we move it to `/icons/$size.jpg` and establish a symlink
3. New icons are generated with ``elgg_create_entity_icons()`` (which is capable of cropping square icons)
4. ``elgg_get_icon_sizes()`` is hookable. File plugin just sets it's own sizes. You can unregister the hook, and all thumbs will be created with site-wide config.

**Comment by hypeJunction** [comment]
I am going to let it sit for a week or so if anyone cares to review and comment, before I put any more time into it. This is last big feature I want to get into 2.2. Please comment now, don't wait until it's ready to merge to only tell me you don't like. Utterly demotivating and frustrating.

File: engine/lib/filestore.php [hunk 6]
```diff
@@ -602,6 +605,356 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = ''
...
+/**
+ * Returns a configuration array of icon sizes
+ *
+ * @param \ElggEntity $entity Entity
+ * @return array
+ */
+function elgg_get_icon_sizes(\ElggEntity $entity = null) {
+	$icon_sizes = _elgg_services()->config->get('icon_sizes');
+
+	if (!$entity) {
```
Comment (hypeJunction) [inline engine/lib/filestore.php]: Should we still use the hook?

File: engine/lib/filestore.php [hunk 7]
```diff
@@ -602,6 +605,356 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = ''
...
+	}
+
+	return elgg_trigger_plugin_hook('entity:icon:sizes', $entity->getType(), ['entity' => $entity], $icon_sizes);
+}
+
+/**
+ * Create entity icons from a source file
+ *
+ * @param \ElggEntity      $entity        Entity to own the icons
+ * @param \ElggFile|string $file          Source image as an \ElggFile object, or path to source file, or form input name
```
Comment (hypeJunction) [inline engine/lib/filestore.php]: Can be null if $entity is ElggFile

File: engine/lib/filestore.php [hunk 8]
```diff
@@ -602,6 +605,356 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = ''
...
+
+	if (!$filename || !is_readable($filename)) {
+		elgg_log('Source file passed to ' . __FUNCTION__ . ' is not readable', 'ERROR');
+		return false;
+	}
+
+	$mimetype = (new \Elgg\Filesystem\MimeTypeDetector())->getType($filename, $default_mimetype);
+	$simpletype = elgg_get_file_simple_type($mimetype);
+
+	if ($simpletype !== 'image') {
```
Comment (hypeJunction) [inline engine/lib/filestore.php]: We can trigger a prepare hook here

**Comment by hypeJunction** [comment]
I think it should be easy to make a service and add tests.

**Comment by hypeJunction** [comment]
I wrote tests #9647 to see what's happening with symlinks on filestore. I think it's safe to use symlinks as a deprecation strategy.

**Comment by hypeJunction** [comment]
Replaced by #9655

**Pull request closed without merging**